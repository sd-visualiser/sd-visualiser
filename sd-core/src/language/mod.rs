use std::{
    fmt::{Debug, Display},
    hash::Hash,
};

use derivative::Derivative;

use crate::{common::Matchable, hypergraph::traits::WithType, prettyprinter::PrettyPrint};

pub mod chil;
#[cfg(not(target_arch = "wasm32"))]
pub mod llvm_ir;
pub mod mlir;
pub mod sd_lang;

pub(crate) fn span_into_str(span: pest::Span) -> &str {
    span.as_str()
}

pub trait GetVar<V> {
    fn var(&self) -> &V;
    fn into_var(self) -> V;
}

impl<V> GetVar<V> for V {
    fn var(&self) -> &V {
        self
    }

    fn into_var(self) -> V {
        self
    }
}

pub trait Fresh {
    fn fresh(number: usize) -> Self;
}

/// `Display` should give the symbolic representation (e.g. "+").
/// `PrettyPrint` should give the textual representation (e.g. "plus").
pub trait Syntax:
    Clone + Eq + Hash + Debug + Send + Sync + Display + Matchable + PrettyPrint
{
}
impl<T: Clone + Eq + Hash + Debug + Send + Sync + Display + Matchable + PrettyPrint> Syntax for T {}

pub enum CF<T: Language + ?Sized> {
    Return,
    Brs(Vec<T::BlockAddr>),
}

pub trait OpInfo<T: Language + ?Sized> {
    fn get_cf(&self) -> Option<CF<T>> {
        None
    }
    fn symbols_used(&self) -> impl Iterator<Item = T::Symbol> {
        std::iter::empty()
    }
    fn sym_name(&self) -> Option<T::Symbol> {
        None
    }
}

pub trait Language {
    type Op: Syntax + OpInfo<Self>;
    type Var: Syntax + Fresh + From<Self::Symbol> + WithType;
    type Addr: Syntax;
    type BlockAddr: Syntax;
    type VarDef: Syntax + GetVar<Self::Var>;
    type Symbol: Syntax + TryFrom<Self::Addr, Error: Debug>;
}

#[derive(Derivative)]
#[derivative(
    Clone(bound = ""),
    Eq(bound = ""),
    PartialEq(bound = ""),
    Hash(bound = ""),
    Debug(bound = ""),
    Default(bound = "")
)]
pub struct Expr<T: Language + ?Sized> {
    pub binds: Vec<Bind<T>>,
    pub values: Vec<Value<T>>,
}

#[derive(Derivative)]
#[derivative(
    Clone(bound = ""),
    Eq(bound = ""),
    PartialEq(bound = ""),
    Hash(bound = ""),
    Debug(bound = "")
)]
pub struct Bind<T: Language + ?Sized> {
    pub defs: Vec<T::VarDef>,
    pub value: Value<T>,
}

#[derive(Derivative)]
#[derivative(
    Clone(bound = ""),
    Eq(bound = ""),
    PartialEq(bound = ""),
    Hash(bound = ""),
    Debug(bound = "")
)]
pub enum Value<T: Language + ?Sized> {
    Variable(T::Var),
    Thunk(Thunk<T>),
    Op { op: T::Op, args: Vec<Value<T>> },
}

#[derive(Derivative)]
#[derivative(
    Clone(bound = ""),
    Eq(bound = ""),
    PartialEq(bound = ""),
    Hash(bound = ""),
    Debug(bound = "")
)]
pub struct Thunk<T: Language + ?Sized> {
    pub addr: T::Addr,
    pub args: Vec<T::VarDef>,
    pub reqs: Vec<T::Var>,
    pub body: Expr<T>,
    pub blocks: Vec<Block<T>>,
}

#[derive(Derivative)]
#[derivative(
    Clone(bound = ""),
    Eq(bound = ""),
    PartialEq(bound = ""),
    Hash(bound = ""),
    Debug(bound = "")
)]
pub struct Block<T: Language + ?Sized> {
    pub addr: T::BlockAddr,
    pub args: Vec<T::VarDef>,
    pub expr: Expr<T>,
}

// Conversions between languages

impl<T: Language> Expr<T> {
    pub fn into<U: Language>(self) -> Expr<U>
    where
        U::Op: From<T::Op>,
        U::Var: From<T::Var>,
        U::Addr: From<T::Addr>,
        U::VarDef: From<T::VarDef>,
        U::BlockAddr: From<T::BlockAddr>,
    {
        Expr {
            binds: self.binds.into_iter().map(Bind::into).collect(),
            values: self.values.into_iter().map(Value::into).collect(),
        }
    }
}

impl<T: Language> Bind<T> {
    pub fn into<U: Language>(self) -> Bind<U>
    where
        U::Op: From<T::Op>,
        U::Var: From<T::Var>,
        U::Addr: From<T::Addr>,
        U::VarDef: From<T::VarDef>,
        U::BlockAddr: From<T::BlockAddr>,
    {
        Bind {
            defs: self.defs.into_iter().map(Into::into).collect(),
            value: self.value.into(),
        }
    }
}

impl<T: Language> Value<T> {
    pub fn into<U: Language>(self) -> Value<U>
    where
        U::Op: From<T::Op>,
        U::Var: From<T::Var>,
        U::Addr: From<T::Addr>,
        U::VarDef: From<T::VarDef>,
        U::BlockAddr: From<T::BlockAddr>,
    {
        match self {
            Self::Variable(var) => Value::Variable(var.into()),
            Self::Thunk(thunk) => Value::Thunk(thunk.into()),
            Self::Op { op, args } => Value::Op {
                op: op.into(),
                args: args.into_iter().map(Value::into).collect(),
            },
        }
    }
}

impl<T: Language> Thunk<T> {
    pub fn into<U: Language>(self) -> Thunk<U>
    where
        U::Op: From<T::Op>,
        U::Var: From<T::Var>,
        U::Addr: From<T::Addr>,
        U::VarDef: From<T::VarDef>,
        U::BlockAddr: From<T::BlockAddr>,
    {
        Thunk {
            addr: self.addr.into(),
            args: self.args.into_iter().map(Into::into).collect(),
            reqs: self.reqs.into_iter().map(Into::into).collect(),
            body: self.body.into(),
            blocks: self.blocks.into_iter().map(Block::into).collect(),
        }
    }
}

impl<T: Language> Block<T> {
    pub fn into<U: Language>(self) -> Block<U>
    where
        U::Op: From<T::Op>,
        U::Var: From<T::Var>,
        U::Addr: From<T::Addr>,
        U::VarDef: From<T::VarDef>,
        U::BlockAddr: From<T::BlockAddr>,
    {
        Block {
            addr: self.addr.into(),
            args: self.args.into_iter().map(Into::into).collect(),
            expr: self.expr.into(),
        }
    }
}

#[cfg(test)]
pub(crate) mod tests {
    use std::{ffi::OsStr, path::Path};

    use serde::Serialize;

    use super::{
        Expr, Language,
        chil::tests::parse_chil,
        llvm_ir::tests::parse_llvm_ir,
        mlir::{
            self,
            internal::{TopLevelItem, tests::parse_mlir},
        },
        sd_lang::tests::parse_sd_lang,
    };
    use crate::{graph::SyntaxHypergraph, hypergraph::petgraph::to_pet, language::llvm_ir::LlvmIr};

    pub trait ExprTest {
        fn free_var_test(&self) -> Box<dyn std::fmt::Debug>;
        fn graph_test(&self, name: &str, lang: &str, sym_name_link: bool) -> anyhow::Result<()>;
    }

    impl<T: Language + 'static> ExprTest for Expr<T>
    where
        <T as Language>::Op: Serialize,
        <T as Language>::BlockAddr: Serialize,
        <T as Language>::Var: Serialize,
        <T as Language>::VarDef: Serialize,
    {
        fn free_var_test(&self) -> Box<dyn std::fmt::Debug> {
            Box::new(self.free_vars(false))
        }

        fn graph_test(&self, name: &str, lang: &str, sym_name_link: bool) -> anyhow::Result<()> {
            let graph: SyntaxHypergraph<T> = self.to_graph(sym_name_link)?;
            let name = if sym_name_link {
                format!("hypergraph_with_sym_{name}.{lang}")
            } else {
                format!("hypergraph_{name}.{lang}")
            };
            insta::assert_ron_snapshot!(name, to_pet(&graph));
            Ok(())
        }
    }

    impl ExprTest for crate::language::llvm_ir::Expr {
        fn free_var_test(&self) -> Box<dyn std::fmt::Debug> {
            Box::new(self.free_vars(false))
        }

        fn graph_test(&self, name: &str, lang: &str, sym_name_link: bool) -> anyhow::Result<()> {
            let graph: SyntaxHypergraph<LlvmIr> = self.to_graph(sym_name_link)?;
            let name = if sym_name_link {
                format!("hypergraph_with_sym_{name}_debug.{lang}")
            } else {
                format!("hypergraph_{name}_debug.{lang}")
            };
            // cannot serialise LLVM IR expressions, so use Debug instead
            insta::assert_debug_snapshot!(name, to_pet(&graph));
            Ok(())
        }
    }

    pub fn parse(raw_path: &str) -> (&str, &str, Box<dyn ExprTest>) {
        let path = Path::new(raw_path);
        match path.extension() {
            Some(ext) if ext == OsStr::new("sd") => {
                let (name, expr) = parse_sd_lang(raw_path);
                ("sd", name, Box::new(expr))
            }
            Some(ext) if ext == OsStr::new("chil") => {
                let (name, expr) = parse_chil(raw_path);
                ("chil", name, Box::new(expr))
            }
            Some(ext) if ext == OsStr::new("ll") => {
                let (name, expr) = parse_llvm_ir(raw_path);
                ("llvm-ir", name, Box::new(expr))
            }
            Some(ext) if ext == OsStr::new("mlir") => {
                let (name, items) = parse_mlir(raw_path);
                let ops: Vec<mlir::internal::Operation> = items
                    .into_iter()
                    .filter_map(|x| match x {
                        TopLevelItem::Operation(y) => Some(y),
                        TopLevelItem::Other(_) => None,
                    })
                    .collect();
                let expr = mlir::Expr::from(ops);
                ("mlir", name, Box::new(expr))
            }
            _ => unreachable!(),
        }
    }
}
