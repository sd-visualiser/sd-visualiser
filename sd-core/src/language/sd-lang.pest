program = _{ SOI ~ expr ~ EOI }

keyword = { "bind" | "in" }

expr     =  { bind* ~ output }
bind     =  { "bind" ~ var_list ~ "=" ~ value ~ "in" }
output   = _{ value | "(" ~ (value ~ ("," ~ value)+)? ~ ")" }
var_list = _{ variable | "(" ~ (variable ~ ("," ~ variable)*) ~ ")" }

thunk = { variable* ~ "." ~ expr }

value = { thunk | variable | op ~ ("(" ~ value ~ ("," ~ value)* ~ ")")? }

op     =  { "plus" | "minus" | "times" | "div" | "rem" | "and" | "or" | "not" | "if" | "eq" | "neq" | "lt" | "leq" | "gt" | "geq" | "app" | "lambda" | "atom" | "deref" | "assign" | "tuple" | "detuple" | "true" | "false" | number }
number = @{ ASCII_DIGIT+ }

variable = @{ !(keyword | op) ~ "_"* ~ ASCII_ALPHA ~ (ASCII_ALPHANUMERIC | "_")* }

COMMENT    = _{ "#" ~ (!"\n" ~ ANY)* }
newline    = _{ "\n" | "\r\n" }
WHITESPACE = _{ " " | "\t" | newline }
