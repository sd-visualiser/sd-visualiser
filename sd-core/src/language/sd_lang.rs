#![allow(clippy::clone_on_copy)]

use std::{
    fmt::{Display, Write},
    str::FromStr,
};

use from_pest::{ConversionError, FromPest, Void};
use pest::iterators::Pairs;
use pest_ast::FromPest;
use pest_derive::Parser;
#[cfg(test)]
use serde::Serialize;

use super::{Fresh, OpInfo, span_into_str};
use crate::{
    common::{Empty, Matchable, Unit},
    hypergraph::traits::{WireType, WithType},
};

pub struct SdLang;

impl super::Language for SdLang {
    type Op = Op;
    type Var = Variable;
    type Addr = Unit;
    type VarDef = Variable;
    type BlockAddr = Empty;
    type Symbol = Empty;
}

impl TryFrom<<SdLang as super::Language>::Addr> for <SdLang as super::Language>::Symbol {
    type Error = &'static str;

    fn try_from(_: <SdLang as super::Language>::Addr) -> Result<Self, Self::Error> {
        Err("no symbols in sd-lang")
    }
}

pub type Expr = super::Expr<SdLang>;
pub type Bind = super::Bind<SdLang>;
pub type Value = super::Value<SdLang>;
pub type Thunk = super::Thunk<SdLang>;

#[derive(Parser)]
#[grammar = "language/sd-lang.pest"]
pub struct SdLangParser;

#[derive(Copy, Clone, Eq, PartialEq, Hash, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub enum Op {
    Plus,
    Minus,
    Times,
    Div,
    Rem,
    And,
    Or,
    Not,
    If,
    Eq,
    Neq,
    Lt,
    Leq,
    Gt,
    Geq,
    App,
    Lambda,
    Atom,
    Deref,
    Assign,
    Tuple,
    Detuple,
    Bool(bool),
    Number(usize),
}

impl Display for Op {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Plus => f.write_char('+'),
            Self::Minus => f.write_char('-'),
            Self::Times => f.write_char('×'),
            Self::Div => f.write_char('/'),
            Self::Rem => f.write_char('%'),
            Self::And => f.write_char('∧'),
            Self::Or => f.write_char('∨'),
            Self::Not => f.write_char('¬'),
            Self::If => f.write_str("if"),
            Self::Eq => f.write_char('='),
            Self::Neq => f.write_char('≠'),
            Self::Lt => f.write_char('<'),
            Self::Leq => f.write_char('≤'),
            Self::Gt => f.write_char('>'),
            Self::Geq => f.write_char('≥'),
            Self::App => f.write_char('@'),
            Self::Lambda => f.write_char('λ'),
            Self::Atom => f.write_char('&'),
            Self::Deref => f.write_char('!'),
            Self::Assign => f.write_str(":="),
            Self::Tuple => f.write_str("()"),
            Self::Detuple => f.write_str(")("),
            Self::Bool(b) => f.write_str(&b.to_string()),
            Self::Number(n) => f.write_str(&n.to_string()),
        }
    }
}

impl FromStr for Op {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "plus" => Ok(Self::Plus),
            "minus" => Ok(Self::Minus),
            "times" => Ok(Self::Times),
            "div" => Ok(Self::Div),
            "rem" => Ok(Self::Rem),
            "and" => Ok(Self::And),
            "or" => Ok(Self::Or),
            "not" => Ok(Self::Not),
            "if" => Ok(Self::If),
            "eq" => Ok(Self::Eq),
            "neq" => Ok(Self::Neq),
            "lt" => Ok(Self::Lt),
            "leq" => Ok(Self::Leq),
            "gt" => Ok(Self::Gt),
            "geq" => Ok(Self::Geq),
            "app" => Ok(Self::App),
            "lambda" => Ok(Self::Lambda),
            "atom" => Ok(Self::Atom),
            "deref" => Ok(Self::Deref),
            "assign" => Ok(Self::Assign),
            "tuple" => Ok(Self::Tuple),
            "detuple" => Ok(Self::Detuple),
            "true" => Ok(Self::Bool(true)),
            "false" => Ok(Self::Bool(false)),
            _ => s.parse().map(Self::Number).map_err(|_err| ()),
        }
    }
}

impl Matchable for Op {
    fn is_match(&self, query: &str) -> bool {
        match self {
            Self::Plus => query == "plus",
            Self::Minus => query == "minus",
            Self::Times => query == "times",
            Self::Div => query == "div",
            Self::Rem => query == "rem",
            Self::And => query == "and",
            Self::Or => query == "or",
            Self::Not => query == "not",
            Self::If => query == "if",
            Self::Eq => query == "eq",
            Self::Neq => query == "neq",
            Self::Lt => query == "lt",
            Self::Leq => query == "leq",
            Self::Gt => query == "gt",
            Self::Geq => query == "geq",
            Self::App => query == "app",
            Self::Lambda => query == "lambda",
            Self::Atom => query == "atom",
            Self::Deref => query == "deref",
            Self::Assign => query == "assign",
            Self::Tuple => query == "tuple",
            Self::Detuple => query == "detuple",
            Self::Bool(b) => b.to_string() == query,
            Self::Number(n) => n.to_string() == query,
        }
    }
}

impl<'pest> FromPest<'pest> for Op {
    type Rule = Rule;
    type FatalError = Void;

    fn from_pest(
        pest: &mut Pairs<'pest, Self::Rule>,
    ) -> Result<Self, ConversionError<Self::FatalError>> {
        let mut clone = pest.clone();
        let pair = clone.next().ok_or(ConversionError::NoMatch)?;
        if pair.as_rule() != Rule::op {
            return Err(ConversionError::NoMatch);
        }
        let op = pair
            .as_str()
            .parse()
            .map_err(|()| ConversionError::NoMatch)?;
        *pest = clone;
        Ok(op)
    }
}

impl OpInfo<SdLang> for Op {}

#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug, FromPest)]
#[cfg_attr(test, derive(Serialize))]
#[pest_ast(rule(Rule::variable))]
pub struct Variable(#[pest_ast(outer(with(span_into_str), with(str::to_string)))] pub String);

impl WithType for Variable {
    fn get_type(&self) -> WireType {
        WireType::Data
    }
}

impl From<Empty> for Variable {
    fn from(value: Empty) -> Self {
        match value {}
    }
}

impl Matchable for Variable {
    fn is_match(&self, query: &str) -> bool {
        self.0 == query
    }
}

impl Display for Variable {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.0)
    }
}

impl Fresh for Variable {
    fn fresh(number: usize) -> Self {
        Self(format!("?{number}"))
    }
}

impl<'pest> FromPest<'pest> for Unit {
    type Rule = Rule;
    type FatalError = Void;

    fn from_pest(_: &mut Pairs<'pest, Rule>) -> Result<Self, ConversionError<Void>> {
        Ok(Unit)
    }
}

// Conversions from pest parse trees

impl<'pest> FromPest<'pest> for Expr {
    type Rule = Rule;
    type FatalError = Void;

    fn from_pest(
        pest: &mut Pairs<'pest, Self::Rule>,
    ) -> Result<Self, ConversionError<Self::FatalError>> {
        let mut clone = pest.clone();
        let pair = clone.next().ok_or(ConversionError::NoMatch)?;
        if pair.as_rule() != Rule::expr {
            return Err(ConversionError::NoMatch);
        }
        let mut inner = pair.into_inner();
        let expr = Expr {
            binds: FromPest::from_pest(&mut inner)?,
            values: FromPest::from_pest(&mut inner)?,
        };
        if inner.clone().next().is_some() {
            return Err(ConversionError::Extraneous {
                current_node: stringify!(Expr),
            });
        }
        *pest = clone;
        Ok(expr)
    }
}

impl<'pest> FromPest<'pest> for Bind {
    type Rule = Rule;
    type FatalError = Void;

    fn from_pest(
        pest: &mut Pairs<'pest, Self::Rule>,
    ) -> Result<Self, ConversionError<Self::FatalError>> {
        let mut clone = pest.clone();
        let pair = clone.next().ok_or(ConversionError::NoMatch)?;
        if pair.as_rule() != Rule::bind {
            return Err(ConversionError::NoMatch);
        }
        let mut inner = pair.into_inner();
        let bind = Bind {
            defs: FromPest::from_pest(&mut inner)?,
            value: FromPest::from_pest(&mut inner)?,
        };
        if inner.next().is_some() {
            return Err(ConversionError::Extraneous {
                current_node: stringify!(Bind),
            });
        }
        *pest = clone;
        Ok(bind)
    }
}

impl<'pest> FromPest<'pest> for Value {
    type Rule = Rule;
    type FatalError = Void;

    fn from_pest(
        pest: &mut Pairs<'pest, Self::Rule>,
    ) -> Result<Self, ConversionError<Self::FatalError>> {
        let mut clone = pest.clone();
        let pair = clone.next().ok_or(ConversionError::NoMatch)?;
        if pair.as_rule() != Rule::value {
            return Err(ConversionError::NoMatch);
        }
        let value = Err(ConversionError::NoMatch)
            .or_else(|_: ConversionError<Void>| {
                let mut inner = pair.clone().into_inner();
                let value = Value::Variable(FromPest::from_pest(&mut inner)?);
                if inner.next().is_some() {
                    return Err(ConversionError::Extraneous {
                        current_node: stringify!(Value),
                    });
                }
                Ok(value)
            })
            .or_else(|_: ConversionError<Void>| {
                let mut inner = pair.clone().into_inner();
                let value = Value::Thunk(FromPest::from_pest(&mut inner)?);
                if inner.next().is_some() {
                    return Err(ConversionError::Extraneous {
                        current_node: stringify!(Value),
                    });
                }
                Ok(value)
            })
            .or_else(|_: ConversionError<Void>| {
                let mut inner = pair.into_inner();
                let value = Value::Op {
                    op: FromPest::from_pest(&mut inner)?,
                    args: FromPest::from_pest(&mut inner)?,
                };
                if inner.next().is_some() {
                    return Err(ConversionError::Extraneous {
                        current_node: stringify!(Value),
                    });
                }
                Ok(value)
            })?;
        *pest = clone;
        Ok(value)
    }
}

impl<'pest> FromPest<'pest> for Thunk {
    type Rule = Rule;
    type FatalError = Void;

    fn from_pest(
        pest: &mut Pairs<'pest, Self::Rule>,
    ) -> Result<Self, ConversionError<Self::FatalError>> {
        let mut clone = pest.clone();
        let pair = clone.next().ok_or(ConversionError::NoMatch)?;
        if pair.as_rule() != Rule::thunk {
            return Err(ConversionError::NoMatch);
        }
        let mut inner = pair.into_inner();
        let addr = FromPest::from_pest(&mut inner)?;
        let args = FromPest::from_pest(&mut inner)?;
        let body = FromPest::from_pest(&mut inner)?;
        let reqs = match inner.next() {
            Some(with) if with.as_rule() == Rule::with => {
                FromPest::from_pest(&mut with.into_inner())?
            }
            _ => Vec::default(),
        };
        let thunk = Thunk {
            addr,
            args,
            reqs,
            body,
            blocks: Vec::default(),
        };
        if inner.next().is_some() {
            return Err(ConversionError::Extraneous {
                current_node: stringify!(Thunk),
            });
        }
        *pest = clone;
        Ok(thunk)
    }
}

pub fn parse_sd_lang(raw_path: &str) -> (&str, Expr) {
    let path = std::path::Path::new(raw_path);
    let program = std::fs::read_to_string(path).unwrap();
    let mut pairs = <SdLangParser as pest::Parser<_>>::parse(Rule::program, &program)
        .unwrap_or_else(|err| {
            panic!(
                "could not parse program {:?}\n{err:?}",
                path.file_stem().unwrap()
            )
        });
    let name = path.file_stem().unwrap().to_str().unwrap();
    let expr = Expr::from_pest(&mut pairs).unwrap();
    (name, expr)
}

#[cfg(test)]
pub(crate) mod tests {

    use dir_test::{Fixture, dir_test};

    use super::Expr;

    #[allow(clippy::needless_pass_by_value)]
    #[dir_test(dir: "$CARGO_MANIFEST_DIR/../examples", glob: "**/*.sd", loader: crate::language::sd_lang::parse_sd_lang, postfix: "check_parse")]
    fn check_parse(fixture: Fixture<(&str, Expr)>) {
        let (_name, _expr) = fixture.content();
    }
}
