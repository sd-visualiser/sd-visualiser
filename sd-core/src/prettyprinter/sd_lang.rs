use pretty::RcDoc;

use super::{PrettyPrint, paran_list};
use crate::language::sd_lang::{Bind, Expr, Op, Thunk, Value, Variable};

impl PrettyPrint for Expr {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        RcDoc::concat(self.binds.iter().map(PrettyPrint::to_doc)).append({
            if self.values.is_empty() {
                RcDoc::text("()")
            } else if self.values.len() == 1 {
                self.values[0].to_doc()
            } else {
                paran_list(&self.values)
            }
        })
    }
}

impl PrettyPrint for Bind {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        RcDoc::text("bind")
            .append(RcDoc::space())
            .append(if self.defs.len() == 1 {
                self.defs[0].to_doc()
            } else {
                paran_list(&self.defs)
            })
            .append(RcDoc::space())
            .append(RcDoc::text("="))
            .append(RcDoc::space())
            .append(self.value.to_doc())
            .append(RcDoc::space())
            .append(RcDoc::text("in"))
            .append(RcDoc::line())
    }
}

impl PrettyPrint for Variable {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        RcDoc::text(&self.0)
    }
}

impl PrettyPrint for Value {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        match self {
            Self::Variable(var) => var.to_doc(),
            Self::Thunk(thunk) => thunk.to_doc(),
            Self::Op { op, args } => {
                if args.is_empty() {
                    op.to_doc()
                } else {
                    op.to_doc().append(paran_list(args))
                }
            }
        }
    }
}

impl PrettyPrint for Op {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        match self {
            Self::Plus => RcDoc::text("plus"),
            Self::Minus => RcDoc::text("minus"),
            Self::Times => RcDoc::text("times"),
            Self::Div => RcDoc::text("div"),
            Self::Rem => RcDoc::text("rem"),
            Self::And => RcDoc::text("and"),
            Self::Or => RcDoc::text("or"),
            Self::Not => RcDoc::text("not"),
            Self::If => RcDoc::text("if"),
            Self::Eq => RcDoc::text("eq"),
            Self::Neq => RcDoc::text("neq"),
            Self::Lt => RcDoc::text("lt"),
            Self::Leq => RcDoc::text("leq"),
            Self::Gt => RcDoc::text("gt"),
            Self::Geq => RcDoc::text("geq"),
            Self::App => RcDoc::text("app"),
            Self::Lambda => RcDoc::text("lambda"),
            Self::Atom => RcDoc::text("atom"),
            Self::Deref => RcDoc::text("deref"),
            Self::Assign => RcDoc::text("assign"),
            Self::Tuple => RcDoc::text("tuple"),
            Self::Detuple => RcDoc::text("detuple"),
            Self::Bool(b) => RcDoc::as_string(b),
            Self::Number(n) => RcDoc::as_string(n),
        }
    }
}

impl PrettyPrint for Thunk {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        RcDoc::intersperse(self.args.iter().map(PrettyPrint::to_doc), RcDoc::space())
            .append(RcDoc::space())
            .append(RcDoc::text("."))
            .append(if self.body.binds.is_empty() {
                RcDoc::space().append(self.body.to_doc())
            } else {
                RcDoc::line()
                    .append(self.body.to_doc())
                    .nest(4)
                    .append(RcDoc::line())
            })
    }
}

#[cfg(test)]
mod tests {
    use dir_test::{Fixture, dir_test};
    use insta::assert_snapshot;

    use crate::{language::sd_lang::Expr, prettyprinter::PrettyPrint};

    #[allow(clippy::needless_pass_by_value)]
    #[dir_test(dir: "$CARGO_MANIFEST_DIR/../examples", glob: "**/*.sd", loader: crate::language::sd_lang::tests::parse_sd_lang, postfix: "pretty_print")]
    fn pretty_print(fixture: Fixture<(&str, Expr)>) {
        let (name, expr) = fixture.content();
        assert_snapshot!(format!("pretty_print_{name}"), expr.to_pretty());
    }
}
