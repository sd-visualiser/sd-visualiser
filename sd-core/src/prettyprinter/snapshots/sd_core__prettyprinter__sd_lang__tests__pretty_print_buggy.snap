---
source: sd-core/src/prettyprinter/sd_lang.rs
expression: expr.to_pretty()
---
lambda(x . lambda(a, y . not(b)))
